pyupgrade/_plugins/defauldict_lambda.py

```diff
@@ -0,0 +1,77 @@
+from __future__ import annotations
+
+import ast
+import functools
+from typing import Any
+from typing import Iterable
+
+from tokenize_rt import Offset
+
+from pyupgrade._ast_helpers import ast_to_offset
+from pyupgrade._ast_helpers import is_name_attr
+from pyupgrade._data import register
+from pyupgrade._data import State
+from pyupgrade._data import TokenFunc
+from pyupgrade._token_helpers import find_and_replace_call
+
+_ITERABLE_BUILTINS = frozenset(('dict', 'list', 'set', 'tuple'))
+
+
+def _lambda_constant_replacement(value: Any) -> str | None:
+    if value == 0:
+        return 'int' if isinstance(value, int) else 'float'
+    if value is None:
+        return ''
+    if value == '':
+        return 'str'
+    return None
+
+
+def _eligible_lambda_replacement(lambda_expr: ast.Lambda) -> str | None:
+    if isinstance(lambda_expr.body, ast.Constant):
+        return _lambda_constant_replacement(lambda_expr.body.value)
+    if isinstance(lambda_expr.body, ast.List) and not lambda_expr.body.elts:
+        return 'list'
+    if isinstance(lambda_expr.body, ast.Tuple) and not lambda_expr.body.elts:
+        return 'tuple'
+    if isinstance(lambda_expr.body, ast.Dict) and not lambda_expr.body.keys:
+        return 'dict'
+    if (
+        isinstance(lambda_expr.body, ast.Call) and
+        isinstance(lambda_expr.body.func, ast.Name) and
+        not lambda_expr.body.args and
+        not lambda_expr.body.keywords and
+        lambda_expr.body.func.id in _ITERABLE_BUILTINS
+    ):
+        return lambda_expr.body.func.id
+    return None
+
+
+@register(ast.Call)
+def visit_Call(
+        state: State,
+        node: ast.Call,
+        parent: ast.AST,
+) -> Iterable[tuple[Offset, TokenFunc]]:
+    if (
+        is_name_attr(
+            node.func,
+            state.from_imports,
+            ('collections',),
+            ('defaultdict',),
+        ) and
+        node.args and
+        isinstance(node.args[0], ast.Lambda) and
+        (replacement := _eligible_lambda_replacement(node.args[0])) is not None
```

Comment:
I haven't seen any walrus usage in pyupgrade but it was convenient here, let me know if you want me to change it thought.

Reply:
please don't use walrus -- I would never